dali/kernels/reduce/reduce_setup_utils.h

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_REDUCE_REDUCE_SETUP_UTILS_H_
+#define DALI_KERNELS_REDUCE_REDUCE_SETUP_UTILS_H_
+
+#include <utility>
+#include <type_traits>
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/span.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/core/util.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+template <typename T>
+constexpr enable_if_t<std::is_integral<T>::value, int> ilog2(T x) {
+  int n = 0;
+  while (x >>= 1)
+    n++;
+  return n;
+}
+
+
+template <typename OutType = uint64_t, typename BitIndices>
+OutType to_bit_mask(const BitIndices &bit_indices) {
```

Comment:
Yes, it will output incorrect mask.
I though about moving this to core/util.h - in that case the thing I'd expect from such a function is _speed_. For the sake of if, I'll not add verification here - especially if I mark this function as `__host__ __device__` - I've had quite painful encounter with `assert` in device code eating up inordinate number of registers.

Reply:
I just wanted to point out, that it's not obvious and the incorrect mask will be an error that's hard to spot later on. How much speed would we sacrifice by verifying value here? In per-cents?

Reply:
If I put an assert there and it makes me change CUDA block size? Perhaps 50%.

Reply:
Anyway, I think you're looking at it as a high level function. It's a _very low level_ function. It's like a glorified bit shift operator - it certainly doesn't throw when you do 1<<100.